tools/afl/src/smessage_sign_roundtrip.c

```diff
@@ -0,0 +1,163 @@
+/*
```

Comment:
why don't enough roundtrip check and use separate decrypt binaries from test files?

Reply:
> why don't enough roundtrip check and use separate decrypt binaries from test files?

I'm sorry, I don't quite understood this. Could you please rephrase?

Reply:
We have `generate/smessage_encrypt_decrypt` binary, which encrypts file and write to file. Plus `src/smessage_encrypt_decrypt`, which decrypts data from a file. And `src/smessage_encrypt_roundtrip` which read file, encrypts and then decrypts encrypted data. The same actions, but by one binary. Why we need first two binaries two and can't leave only `smessage_encrypt_roundtrip`?

Reply:
> Why we need first two binaries two and can't leave only smessage_encrypt_roundtrip?

Oh, now I understood your concerns! Okay...

There are two binaries in order to fuzz the encryption and decryption code paths separately.

The way AFL works is that it mutates only initial inputs. It does not mess with the control flow of an executing process (it does monitor it though, to see the effect of the changes it makes to the input). Therefore, we have two binaries that exercise the following flows:

- mutated(plaintext, keys) => `encrypt()` => ciphertext => `decrypt()` => expect same plaintext
- mutated(ciphertext, keys) => `decrypt()` => expect expected plaintext

Note that in the first case the ciphertext is *not* mutated by AFL. Therefore, `decrypt()` can only be fuzzed via that path only if the input it mutated in such a particular way that encryption succeeds but produces a ciphertext so broken that will trigger `decrypt()` to fail. Unlikely, but we look for that too. More likely, though, is for it to break from come corruption, which is exactly what the second flow tests.

Also note that AFL looks for crashes. If we want to look for an invalid result being returned, we have to make it into a crash. We do so by using `abort()`. Effectively, we're looking at the following flows:

The **roundtrip** binaries:

- mutated(plaintext, keys) => 💥 

  So broken that we can't even parse the input data. Those are ignored.

- mutated(plaintext, keys) => `encrypt()` => 💥

  Invalid plaintext or key triggered a crash (*not* detected and reported an error).

- mutated(plaintext, keys) => `encrypt()` => ciphertext => `decrypt()` => 💥

  That rare case from above. Plaintext or keys were broken just the right way for encryption to actually succeed, but produce some weird ciphertext which causes decryption to crash.

- mutated(plaintext, keys) => `encrypt()` => ciphertext => `decrypt()` => not plaintext => 💥

  Even rarer, probably impossible case where decryption also succeeds, but somehow produces a different plaintext without Themis noticing the discrepancy.

The **encrypt/decrypt** binaries:

- mutated(ciphertext, keys) => 💥

  Again, so broken that we can't even parse the input data. Ignore.

- mutated(ciphertext, keys) => `decrypt()` => 💥

  Looking for crashes caused by encrypted data to be corrupted in a creative way.

- mutated(ciphertext, keys) => `decrypt()` => not expected plaintext => 💥

  Super-rare, impossible case where you can corrupt the data, have it successfully decrypted, but the modification is not noticed by integrity checks of Themis.